### PR TITLE
Call `set_epoch` for distributed batch samplers

### LINKED ISSUE #13316: set_epoch not called for BatchSampler
## 🐛 Bug

Lightning takes care of calling set_epoch for custom DataLoader Samplers [here](https://github.com/Lightning-AI/lightning/blob/759e89df21684277da996dd25781dee7f21e1b1c/src/pytorch_lightning/loops/fit_loop.py#L239). However, the DataLoader might use a custom BatchSampler instead of a Sampler. Lightning does not call set_epoch for custom BatchSamplers. Calling set_epoch is important for proper seeding in distributed environments. 

### Expected behavior

Lightning should call set_epoch for BatchSamplers to match its behavior for Samplers. 
Can use the DataLoader's [index_sampler](https://github.com/pytorch/pytorch/blob/61a1eef7fc44a74591abae2eff1547efc45e915c/torch/utils/data/dataloader.py#L445) property to retrieve the Sampler or BatchSampler that is actually being used by the DataLoader, or more simply, call set_epoch for both Sampler and BatchSampler.

### Additional context

A workaround is to use a Callback such as 

```
class SetBatchSamplerEpoch(Callback):
    """ sets the epoch for batch sampler before dataloader iterator is initialized every training epoch """
    def __init__(self):
        super().__init__()

    @staticmethod
    def set_batch_sampler_epoch(dataloader, epoch):
        if callable(getattr(dataloader.batch_sampler, "set_epoch", None)):
            # print("[RANK {}] Setting batch_sampler epoch to {}".format(os.getenv("LOCAL_RANK", '0'), epoch))
            dataloader.batch_sampler.set_epoch(epoch)

    def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        if trainer.train_dataloader is not None:
            if isinstance(trainer.train_dataloader, CombinedLoader):
                # is the train_dataloader always wrapped in a CombinedLoader at this point?
                apply_to_collection(data=trainer.train_dataloader.loaders,
                                    dtype=DataLoader,
                                    function=self.set_batch_sampler_epoch,
                                    epoch=trainer.current_epoch)

            elif isinstance(trainer.train_dataloader, DataLoader):
                # just in case the train_dataloader is not wrapped in a CombinedLoader
                self.set_batch_sampler_epoch(trainer.train_dataloader, trainer.current_epoch)

            else:
                raise TypeError(f"Unexpected type of trainer.train_dataloader: {type(trainer.train_dataloader)}")
```

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #13316

- The loops now call either `dataloader.sampler.set_epoch` or `dataloader.batch_sampler.set_epoch` if applicable
- Moved duplicated code to a utility function
- Improved tests and fixed description

### Does your PR introduce any breaking changes? If yes, please list them.

Yes (reproducibility). If you relied on distributed batch samplers before and didn't know that Lightning didnt handle the shuffling, you will get different results. This is unavoidable.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/Lightning-AI/lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj @carmocca @ananthsub

### CODE REVIEW

File: src/pytorch_lightning/loops/utilities.py [hunk 1]
```diff
@@ -220,3 +221,16 @@ def _reset_progress(loop: Loop) -> None:
 def _v1_8_output_format(fx: Callable) -> bool:
     parameters = inspect.signature(fx).parameters
     return "new_format" in parameters and parameters["new_format"].default is True
+
+
+def _set_sampler_epoch(dataloader: DataLoader, epoch: int) -> None:
```
Comment (awaelchli) [inline src/pytorch_lightning/loops/utilities.py:226]: added to loops utilities because it is used in loops, but could also be categorized as data utilities.

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/loops/utilities.py [hunk 2]
```diff
@@ -220,3 +221,16 @@ def _reset_progress(loop: Loop) -> None:
...
+    """Calls the ``set_epoch`` method on either the sampler or the batch sampler of the given dataloader.
+
+    Every PyTorch dataloader has either a sampler or a batch sampler, and if it is wrapped by a
+    :class:`~torch.utils.data.DistributedSampler`, ``set_epoch`` must be called at the beginning of every epoch to
+    ensure shuffling applies a new ordering. This has no effect if shuffling is off.
```
Comment (rohitgr7) [inline src/pytorch_lightning/loops/utilities.py:231]:
```suggestion
    :class:`~torch.utils.data.distributed.DistributedSampler`, ``set_epoch`` must be called at the beginning
    of every epoch to ensure shuffling applies a new ordering. This has no effect if shuffling is off.
```

**Review by rohitgr7 — APPROVED** [review]
nit
